File: src/main/java/com/manifestors/shinrai/client/utils/Discord/RPCEngine.java
package com.manifestors.shinrai.client.utils.Discord;

import com.manifestors.shinrai.client.Shinrai;
import net.arikia.dev.drpc.DiscordRPC;
import net.arikia.dev.drpc.DiscordRichPresence;
import net.arikia.dev.drpc.DiscordEventHandlers;

public class RPCEngine {
    private static String SelectedOperatingSystem;
    public static void getOperatingSystem(){
        String os = System.getProperty("os.name").toLowerCase();
        if (os.contains("win")) {
            SelectedOperatingSystem = "win";
        } else if (os.contains("mac")) {
            SelectedOperatingSystem = "mac";
        } else if (os.contains("nix") || os.contains("nux") || os.contains("aix")) {
            SelectedOperatingSystem = "lin";
        } else {
            SelectedOperatingSystem = "organizationlogo";
        }

    }
    public static void StartRPC() {
        DiscordEventHandlers handlers = new DiscordEventHandlers.Builder().build();
        DiscordRPC.discordInitialize("849293997771456532", handlers, true);

        DiscordRichPresence presence = new DiscordRichPresence.Builder("On Main Menu")
                .setDetails("No Activity")
                .setBigImage("logo","Shinrai Minecraft Client")
                .setSmallImage(SelectedOperatingSystem,SelectedOperatingSystem)
                .setStartTimestamps(System.currentTimeMillis())
                .build();

        DiscordRPC.discordUpdatePresence(presence);


        new Thread(() -> {
            while (!Thread.currentThread().isInterrupted()) {
                DiscordRPC.discordRunCallbacks();
                try {
                    Thread.sleep(2000);
                } catch (InterruptedException e) {
                    break;
                }
            }
        }, "Discord-RPC-Callback-Thread").start();
    }

    public static void WorlsRPC(String WorldName) {
        DiscordEventHandlers handlers = new DiscordEventHandlers.Builder().build();
        DiscordRPC.discordInitialize("849293997771456532", handlers, true);

        DiscordRichPresence presence = new DiscordRichPresence.Builder("Playing in a singleplayer world")
                .setDetails(WorldName + "("+ Shinrai.NAME + " " + Shinrai.VERSION +")")
                .setBigImage("shinrailogominecraft",WorldName)
                .setSmallImage(SelectedOperatingSystem,SelectedOperatingSystem)
                .setStartTimestamps(System.currentTimeMillis())
                .build();

        DiscordRPC.discordUpdatePresence(presence);


        new Thread(() -> {
            while (!Thread.currentThread().isInterrupted()) {
                DiscordRPC.discordRunCallbacks();
                try {
                    Thread.sleep(2000);
                } catch (InterruptedException e) {
                    break;
                }
            }
        }, "Discord-RPC-Callback-Thread").start();
    }
}

File: src/main/kotlin/com/manifestors/shinrai/client/utils/discord/RPCEngine.kt
package com.manifestors.shinrai.client.utils.discord

import com.manifestors.shinrai.client.Shinrai.moduleManager
import com.manifestors.shinrai.client.Shinrai.fullVersion
import com.manifestors.shinrai.client.ui.title.ShinraiTitleScreen
import com.manifestors.shinrai.client.utils.MinecraftInstance
import net.arikia.dev.drpc.DiscordEventHandlers
import net.arikia.dev.drpc.DiscordRPC
import net.arikia.dev.drpc.DiscordRichPresence
import net.minecraft.client.gui.screen.multiplayer.MultiplayerScreen
import net.minecraft.client.gui.screen.world.SelectWorldScreen
import java.util.*

object RPCEngine : MinecraftInstance {

    private var isRPCRunning = false
    private var gameStartedAt = 0L

    val operatingSystemData: Array<String>
        get() {
            val os = System.getProperty("os.name").lowercase(Locale.getDefault())
            if (os.contains("win")) {
                return arrayOf("win", "Windows")
            } else if (os.contains("mac")) {
                return arrayOf("mac", "macOS")
            } else if (os.contains("nix") || os.contains("nux") || os.contains("aix")) {
                return arrayOf("lin", "Linux")
            }

            return arrayOf("organizationlogo", "Unknown OS")
        }

    fun startRPC() {
        val handlers = DiscordEventHandlers.Builder().build()
        DiscordRPC.discordInitialize("849293997771456532", handlers, true)
        isRPCRunning = true
        gameStartedAt = System.currentTimeMillis() / 1000
        Thread(Runnable {
            while (isRPCRunning) {
                runRPC()
                try {
                    Thread.sleep(500)
                } catch (ignored: InterruptedException) {
                }
                DiscordRPC.discordRunCallbacks()
            }
        }, "Discord-RPC-Callback-Thread").start()
    }

    fun shutdownRPC() {
        isRPCRunning = false
        DiscordRPC.discordShutdown()
    }

    fun runRPC() {
        val totalledModules =
            "Enabled " + moduleManager.enabledModules.size + " of " + moduleManager
                .modules.size + " modules"

        val presence = DiscordRichPresence.Builder(state)
            .setDetails(totalledModules)
            .setBigImage("logo", fullVersion)
            .setSmallImage(operatingSystemData[0], operatingSystemData[1])
            .setStartTimestamps(gameStartedAt)
            .build()

        DiscordRPC.discordUpdatePresence(presence)
    }

    private val state: String
        get() {
            if (mc.currentScreen is ShinraiTitleScreen) return "In main menu"
            if (mc.currentScreen is SelectWorldScreen) return "Selecting a world"
            if (mc.currentScreen is MultiplayerScreen) return "Looking for servers"
            if (mc.player != null) {
                val server = mc.server
                return if (mc.isIntegratedServerRunning || server == null) "Playing in a singleplayer world" else "Playing on " + server.serverIp
            }

            return "No activity"
        }
}
